From 0a1e2b766a5c8e65796156d1cec68610fb5f93ed Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Fri, 8 Feb 2019 10:56:29 +0530
Subject: [PATCH 0963/1051] soc: octeontx2-rm: Defer probe if discovery id is
 not setup

Check if discovery ID is setup by AF, if not, defer
driver probe until AF is up.

Change-Id: I948b5c0a28f7444a079aaa734ef8337cbe0d58b7
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/3697
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/soc/marvell/octeontx2-rm/otx2_rm.c | 23 ++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
index a398603a4319..29ed1cd899de 100644
--- a/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
+++ b/drivers/soc/marvell/octeontx2-rm/otx2_rm.c
@@ -1085,6 +1085,25 @@ static int vf_sysfs_create(struct pci_dev *pdev)
 	return err;
 }
 
+static int rm_check_pf_usable(struct rm_dev *rm)
+{
+	u64 rev;
+
+	rev = rm_read64(rm, BLKADDR_RVUM, 0,
+			RVU_PF_BLOCK_ADDRX_DISC(BLKADDR_RVUM));
+	rev = (rev >> 12) & 0xFF;
+	/* Check if AF has setup revision for RVUM block,
+	 * otherwise this driver probe should be deferred
+	 * until AF driver comes up.
+	 */
+	if (!rev) {
+		dev_warn(&rm->pdev->dev,
+			 "AF is not initialized, deferring probe\n");
+		return -EPROBE_DEFER;
+	}
+	return 0;
+}
+
 static int rm_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 {
 	struct device *dev = &pdev->dev;
@@ -1138,6 +1157,10 @@ static int rm_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 		goto set_mask_failed;
 	}
 
+	err = rm_check_pf_usable(rm);
+	if (err)
+		goto set_mask_failed;
+
 	/* Map PF-AF mailbox memory */
 	rm->af_mbx_base = ioremap_wc(pci_resource_start(pdev, PCI_MBOX_BAR_NUM),
 				     pci_resource_len(pdev, PCI_MBOX_BAR_NUM));
-- 
2.17.1

